### PR TITLE
Update async to add async/await to reduce

### PR DESCRIPTION
**Pull request by stevekanter**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/caolan/async/blob/master/lib/reduce.js#L133
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## async ([<kbd>unpkg</kbd>](https://unpkg.com/browse/async@latest/))
was missing the following properties:
1. groupBy
1. groupByLimit
1. groupBySeries
1. tryEach
1. flatMap

<details>
<summary>as well as these 2 other properties...</summary>
<p>flatMapLimit, flatMapSeries</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0017ed5b3ab853a3e7d3a95a668aaaf97ee4f472
</p>

File: types/async/index.d.ts [hunk 1]
```diff
@@ -277,7 +277,8 @@ export const selectLimit: typeof filterLimit;
 export const reject: typeof filter;
 export const rejectSeries: typeof filter;
 export const rejectLimit: typeof filterLimit;
-export function reduce<T, R, E = Error>(arr: T[] | IterableIterator<T>, memo: R, iterator: AsyncMemoIterator<T, R, E>, callback?: AsyncResultCallback<R, E>): void;
+export function reduce<T, R, E = Error>(arr: T[] | IterableIterator<T>, memo: R, iterator: AsyncMemoIterator<T, R, E>, callback: AsyncResultCallback<R, E>): void;
```
Comment (apnsngr) [inline types/async/index.d.ts:280]:
```suggestion
export function reduce<T, R, E = Error>(arr: IterableCollection<T>, memo: R, iterator: AsyncMemoIterator<T, R, E>, callback: AsyncResultCallback<R, E>): void;
```

File: types/async/index.d.ts [hunk 2]
```diff
@@ -277,7 +277,8 @@ export const selectLimit: typeof filterLimit;
...
+export function reduce<T, R, E = Error>(arr: T[] | IterableIterator<T>, memo: R, iterator: AsyncMemoIterator<T, R, E>): Promise<R>;
```
Comment (apnsngr) [inline types/async/index.d.ts:281]:
```suggestion
export function reduce<T, R, E = Error>(arr: IterableCollection<T>, memo: R, iterator: AsyncMemoIterator<T, R, E>): Promise<R>;
```

Comment (apnsngr) [inline types/async/index.d.ts:281]: Please update the [test/index.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/async/test/index.ts) file to include an example that uses this type signature. This file should fail type checking without your patch and pass type checking with your patch.

**Review by apnsngr — CHANGES_REQUESTED** [review]
Thank you for submitting this update! I have requested a couple of quick type clean ups and for a unit test to be added. Let me know if you need more detailed instructions for adding the unit test.

**Pull request closed without merging**